crates/cli/src/opts/build/utils.rs

```diff
@@ -76,13 +79,115 @@ pub fn configure_pcx(
 /// - Source files are added if `add_source_file` is set
+pub fn configure_pcx_from_compile_output(
+    pcx: &mut ParsingContext<'_>,
+    config: &Config,
+    output: &ProjectCompileOutput,
+    target_paths: Option<&[PathBuf]>,
+) -> Result<()> {
+    // If targets are specified, find the max version among those files and their dependencies.
+    let (version, source_paths): (Version, Vec<PathBuf>) = if let Some(targets) = target_paths {
+        let mut scope = HashSet::new();
+        let mut queue: VecDeque<PathBuf> = targets
+            .iter()
+            .filter_map(|path| {
+                if let Ok(full_path) = dunce::canonicalize(path)
+                    && output
+                        .graph()
+                        .get_parsed_source(full_path.as_path())
+                        .is_some_and(|ps| matches!(ps, MultiCompilerParsedSource::Solc(..)))
+                {
+                    Some(full_path)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        while let Some(path) = queue.pop_front() {
+            if scope.insert(path.to_path_buf()) {
+                for import in output.graph().imports(path.as_path()) {
+                    queue.push_back(import.to_path_buf());
+                }
+            }
+        }
+
+        let version = output
+            .output()
+            .sources
+            .sources_with_version()
+            .filter_map(|(p, _, v)| {
+                if let Ok(full_path) = dunce::canonicalize(p)
+                    && scope.contains(&full_path)
+                {
+                    Some(v)
+                } else {
+                    None
+                }
+            })
+            .min()
+            .cloned()
+            .ok_or_else(|| eyre::eyre!("no Solidity sources"))?;
+
+        (version, scope.into_iter().collect())
+    }
+    // Otherwise, find the latest version among all compiled files.
+    else {
+        let (version, paths) = output
+            .output()
+            .sources
+            .sources_with_version()
+            .filter(|(path, _, _)| {
+                // Only process Solidity files.
+                output
+                    .graph()
+                    .get_parsed_source(path)
+                    .is_some_and(|ps| matches!(ps, MultiCompilerParsedSource::Solc(..)))
+            })
+            .fold((Version::new(0, 0, 0), Vec::new()), |(max_v, mut paths), (path, _, version)| {
+                match version.cmp(&max_v) {
+                    core::cmp::Ordering::Less => (version.clone(), vec![path]),
+                    core::cmp::Ordering::Equal => {
+                        paths.push(path);
+                        (max_v, paths)
+                    }
+                    core::cmp::Ordering::Greater => (max_v, paths),
+                }
+            });
+
+        (version, paths.into_iter().filter_map(|p| dunce::canonicalize(p).ok()).collect())
```

Comment:
this is a bit hard to follow, can you rewrite with a single for loop, also combining this second iter here?